Source/src/WixSharp.UI/ManagedUI/Forms/ProgressDialog.cs

```diff
@@ -108,54 +105,37 @@ public override MessageResult ProcessMessage(InstallMessage messageType, Record
-                                   messageRecord[1]    "InstallFiles"
-                                   messageRecord[2]    "Copying new files"
-                                   messageRecord[3]    "File: [1],  Directory: [9],  Size: [6]"
-
-                                   messageRecord[0]    "Action 23:15:21: [1]. [2]"
-                                   messageRecord[1]    "RegisterUser"
-                                   messageRecord[2]    "Registering user"
-                                   messageRecord[3]    "[1]"
-
-                                */
-                                if (messageRecord.FieldCount >= 3)
-                                {
-                                    message = messageRecord[2].ToString();
-                                }
-                            }
-                            else
+                            /*
+                            messageRecord[0] - is reserved for FormatString value
+                            messageRecord[2] unconditionally contains the string to display
+
+                            Examples:
+
+                               messageRecord[0]    "Action 23:14:50: [1]. [2]"
+                               messageRecord[1]    "InstallFiles"
+                               messageRecord[2]    "Copying new files"
+                               messageRecord[3]    "File: [1],  Directory: [9],  Size: [6]"
+
+                               messageRecord[0]    "Action 23:15:21: [1]. [2]"
+                               messageRecord[1]    "RegisterUser"
+                               messageRecord[2]    "Registering user"
+                               messageRecord[3]    "[1]"
+
+                            */
+                            string message = null;
+                            if (messageRecord.FieldCount >= 3)
                             {
-                                message = messageRecord.FormatString;
-                                if (message.IsNotEmpty())
-                                {
-                                    for (int i = 1; i < messageRecord.FieldCount; i++)
-                                    {
-                                        message = message.Replace("[" + i + "]", messageRecord[i].ToString());
-                                    }
-                                }
-                                else
-                                {
-                                    message = messageRecord[messageRecord.FieldCount - 1].ToString();
-                                }
+                                message = messageRecord[2].ToString();
                             }
 
                             if (message.IsNotEmpty())
-                                currentAction.Text = currentActionLabel.Text + " " + message;
+                                currentAction.Text = string.Format("{0} {1}", currentActionLabel.Text, message);
```

Comment:
Are you sure? do we need to be backwards compatible to .net 3.5? which is why I didn't use string interpolation

Reply:
Yep. good point. overlooked it. 

Then to be honest I do prefer concatenation in this case. `Format` heart my eyes :)
Affect readability, does not improve anything.

If you reall-really dislike concatenation then proceed with format extension method.:

```C#
"{0} {1}".FormatWith(currentActionLabel.Text, message);
```

Reply:
`String.Format` is compatible with I18n should that be required at a later date (and also suppresses the SQ issue). I think in general the reason to avoid string contamination is the performance, and semantic concerns see: https://dlang.org/spec/arrays.html#array-concatenation

Reply:
done

Reply:
> I think in general the reason to avoid string concatenation is the performance

I expected that :). This topic is as old as .NET itself.

You are absolutely right. This is the reason why concatenation is usually avoided. 
Ironically in the execution context like here concatenation is actually the fastest option. Followed by `StringBuilder` and then by `Format`. Concatenation has other drawback but not the performance in this case.

Though even if it was the slowest option I would not push for its elimination as it is used in the performance non-critical routine. We don't want go "premature optimization" path :)

As a selection criteria I would stick with readability. If you ask me, string interpolation is a champion here. Followed by concatenation, then `FormatWith`, then `Format` and then `StringBuilder`.

Anyway we do have a good compromise. `FormatWith` is readable enough. SQ will be happy. And Neal has done it already...